### ISSUE TITLE
SkinnedMeshes ignore the frustumCulled property

### ISSUE THREAD

**Issue by hybridherbst** [mrdoob/three.js#24018]
**Describe the bug**

Seems the calculations for culling away animated skinned meshes aren't fully correct; in this example file, the mesh disappears a lot randomly, especially in the later part of the file.

**To Reproduce**

_In model-viewer_
1. Go to https://modelviewer.dev/editor
3. Unpack + Drop [SkinningCullingBug.zip](https://github.com/mrdoob/three.js/files/8643711/SkinningCullingBug.zip)
4. Select the 🖊️ pen Icon
5. Toggle "Play" off
6. Scrub to 250
7. Toggle "Play" back on
8. Watch the character pop in and out of view

_In three.js/editor_
1. Go to [https://threejs.org/editor/](https://threejs.org/editor/)
3. Unpack + Drop [SkinningCullingBug.zip](https://github.com/mrdoob/three.js/files/8643711/SkinningCullingBug.zip)
9. Add a Directional Light
10. Select "SkinningCullingBug.glb", set animation time scale to 10 (the bug is more apparent in later parts of the animation when the character is scaled up)
11. Press Play
12. Watch the animation, you can set the timescale back to 1 (or lower) when the Character becomes bigger to see the issue better

![ezgif com-gif-maker (9)](https://user-images.githubusercontent.com/2693840/167219944-cf8d4672-49f9-4a16-9d44-db55375b9178.gif)

***Live example***

Here's a video:
https://user-images.githubusercontent.com/2693840/167219369-588fe804-8fab-4c42-a5b4-79d4d38250dd.mp4

If you drop the same file into https://sandbox.babylonjs.com/, the character is visible all the time (as expected).

**Expected behavior**

Character is always visible

**Platform:**

 - Device: [Desktop, Mobile]
 - OS: [Windows]
 - Browser: [Chrome]
 - Three.js version: [latest on editor]

**Comment by donmccurdy** [comment]
Unfortunately a long-standing issue: https://github.com/mrdoob/three.js/issues/14499

Current workarounds are to disable frustum culling for SkinnedMesh objects or to upscale the bounding box of the geometry by some factor.

**Comment by hybridherbst** [comment]
Oh no :(

I don't have control over threejs in most cases, e.g. when using model-viewer or other existing viewers.

For reference, @mrdoob had mentioned here

- https://github.com/google/model-viewer/issues/3189#issuecomment-1035603330

that he came up with an efficient solution; but I'm not sure that's already implemented anywhere?

**Comment by Mugen87** [comment]
Marking this as duplicate of #14499.

**Comment by mrdoob** [comment]
@hybridherbst Maybe you could make a PR in the modelviewer repo disabling frustum culling on `SkinnedMesh`?

**Comment by hybridherbst** [comment]
I actually looked into that but found that it's supposed to already be off!

- https://github.com/google/model-viewer/blob/f085bc1d2f46b851b95656713ec0b8ae9bbd9649/packages/model-viewer/src/three-components/gltf-instance/ModelViewerGLTFInstance.ts#L57

@elalish do you maybe have an idea what's going on? I can open an issue on model-viewer if you want of course. There's even a unit test that ensures frustum culling is off.

**Comment by hybridherbst** [comment]
Turns out turning off frustumCulled in the three.js/editor also doesn't affect culling. Maybe there's another issue here?
![image](https://user-images.githubusercontent.com/2693840/167403542-187ce0ca-4c96-4f3c-b3ae-f82a68dc5bea.png)

**Comment by hybridherbst** [comment]
@Mugen87 can I rename the issue to "SkinnedMeshes ignore the frustumCulled property" and could you reopen it?

**Comment by Mugen87** [comment]
Done!

**Comment by Mugen87** [comment]
Can you please set the `frustumCulled` property to `false` for the skinned mesh itself and for all its descendants? 

If an object gets frustum culled, it does only affect the object itself but not its children.

**Comment by hybridherbst** [comment]
Can there be invisible descendants for a skinned mesh? In the above screenshot from the three.js editor that's the leaf node, there's no further children as far as I can see.

**Comment by hybridherbst** [comment]
Updated the original post with a truncated animation that only shows the part with the biggest issues. Seems to be related to scaling up all bones of a SkinnedMesh.

**Comment by donmccurdy** [comment]
Just to clarify, are you describing one of these two scenarios? Or something else?

**(A)**

- Object3D
    - Bone1
    - Bone2
- Object3D (scale=0)
    - SkinnedMesh

**(B)**

- Object3D (scale=0)
    - Bone1
    - Bone2
- Object3D
    - SkinnedMesh

**Comment by hybridherbst** [comment]
**Scenario A** it is - usually the object scale doesn't affect the final result for skinned meshes, since the bones have the ultimate say, but for scale=0 I can see how that _might_ be undefined.

For Scenario B, it would certainly be expected that the result isn't visible (but the skeleton helper / bones would also have scale 0 then).

**Comment by hybridherbst** [comment]
Here's a minimal repro model:
[BentCylinder_Scales.zip](https://github.com/mrdoob/three.js/files/8653687/BentCylinder_Scales.zip)

From left to right:
- SkinnedMesh is scaled to 1 (in this case, 100x smaller than the bind pose)
- SkinnedMesh is scaled to 100 (same as in bind pose)
- SkinnedMesh is scaled to 0.0001
- SkinnedMesh is scaled to 0
- SkinnedMesh has a position shift=10000

**three.js/model-viewer:**
![image](https://user-images.githubusercontent.com/2693840/167459215-e553a289-771a-4e08-8efa-ec698c5ee627.png)

**Babylon:**
![image](https://user-images.githubusercontent.com/2693840/167459343-d1f9e1e2-76aa-4bf6-8b34-f49f08d70b3d.png)

**Unity:**
![image](https://user-images.githubusercontent.com/2693840/167459260-d993dc26-78e9-4233-b6ac-54ee868c0e93.png)

**Comment by elalish** [comment]
This has the look of a numerical problem; maybe generating NaNs somewhere?

**Comment by hybridherbst** [comment]
Yep - I think the questions here would be "is this behaviour defined in glTF" and/or "should three have the same result as Babylon and Unity". Answer might be "this is too much of an edge case" :)

**Comment by donmccurdy** [comment]
glTF Validator should warn about these cases with `NODE_SKINNED_MESH_NON_ROOT`. While the glTF spec says the transform of a skinned mesh must be ignored, skinning is implemented very differently across engines (perhaps Unity especially...) and it's hard to guarantee that a non-invertible matrix — or reaching precision limits — is not going to have side effects.

If I remember correctly, Unity keeps a skinned mesh out of the scene hierarchy entirely, and the transform is completely ignored. three.js does at least compute transforms all the way down, a SkinnedMesh does not get special treatment there, and under some [bind modes](https://threejs.org/docs/?q=skinnedmesh#api/en/objects/SkinnedMesh.bindMode) the SkinnedMesh world matrix can be optionally applied. 

GLTFLoader does not change the default, `bindMode = "attached"`. That's unintuitive to me, "detached" would _sound_ like what the glTF specification calls for, but it has otherwise behaved correctly under this mode w/r/t parent transforms.

If there's a simple change in GLTFLoader (e.g. bindMode -> "detached") or WebGLRenderer (e.g. upload identity world matrix for a "detached" SkinnedMesh?) I think we could make that change, I don't think we could make more complex changes for this case.

/cc #17926 I thought there was code somewhere that simply prevented objects with zero scale from rendering, as a workaround for certain cases like flip-board visibility animation. But I can't find that code now, perhaps I was incorrect.

**Comment by hybridherbst** [comment]
Interesting, the spec even explicitly calls out that implementations MUST not hide a SkinnedMesh if all its axes are scaled to zero:
https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#:~:text=When%20the%20scale,to%20zero%20simultaneously.

> **Implementation Note**
When the scale is zero on all three axes (by node transform or by animated scale), implementations are free to optimize away rendering of the node’s mesh, and all of the node’s children’s meshes. This provides a mechanism to animate visibility. Skinned meshes must not use this optimization unless all of the joints in the skin are scaled to zero simultaneously.
